### DISCUSSION TITLE
CRUD operations on ECsql and other questions

### DISCUSSION THREAD

**Discussion by Nhillus** [iTwin/itwinjs-core#6594] [Q&A] [answered]
Hello, I'm new to the community and have been exploring the iTwin platform and iTwin.js documentation to understand how ECSQL facilitates interactions with iModels, I've observed a lot on read operations in the examples and proyects.

 so my first question: Is it possible to perform UPDATE operations, or to use commands equivalent to "DESC" for exploring table schemas in ECSQL? Based on my readings, it seems these operations might not be supported, but I would like to get confirmation on this.

Additionally, I have a question concerning the concept of working locally as defined by iTwin.js. I understand that it's possible to download iModels and interact with them on a local machine. When iTwin.js docs mentions the ability to download the iModel and work locally, does this refer solely to collaboration within a team, or could it also mean making modifications—such as updating a comment on a piece of functional equipment—and later uploading these changes back to the Imodel inside the Itwin in the iTwin platform?

I appreciate any insights, or guidance Thank you very much in advance!

**Comment by pmconne** [comment, marked as answer]
You cannot change the contents of an iModel directly using ECSql statements like INSERT, UPDATE, and DELETE. You must instead use the TypeScript APIs provided by @itwin/core-backend like [updateElement](https://www.itwinjs.org/reference/core-backend/imodels/imodeldb/imodeldb.elements/updateelement/). For [interactive editing](https://www.itwinjs.org/learning/interactiveediting/) workflows you would typically rely on [tools](https://www.itwinjs.org/learning/frontend/primitivetools/) to perform those operations.

Interacting with an iModel locally involves downloading a [briefcase](https://www.itwinjs.org/reference/core-backend/imodels/briefcasedb/). You can then push your own local changes and pull other people's changes.

HTH. If you have further questions, please specify what you are trying to accomplish.

**Reply by Nhillus** [reply]
Hi @pmconne, thanks for your quick response and for clarifying my queries. I'm trying to make changes to a sample iModel, specifically aiming to modify one of the properties of what I believe is called the "Bay Town Process Plant" sample. My goal is to update just one of the empty properties.

In my initial post, I mentioned wanting to update the comments appearing in the properties, as shown in the photo I've just uploaded. Here's the image link again for reference:

![image](https://github.com/iTwin/itwinjs-core/assets/46725871/5fde465b-4af3-4e48-a511-95aec4345b6a)

Regarding my attempt to work with a briefcase, I was trying to use the downloadBriefcase method to download the iModel I mentioned, but I've encountered some issues. Maybe i should follow your suggest of using updateElement or triying that doc of interactive editing?

Thank you again for your support.

**Reply by pmconne** [reply]
> I've encountered some issues.

Such as?

**Reply by Nhillus** [reply]
the error im getting when triying to use downloadBriefcase  its : No BackendHubAccess supplied in IModelHostOptions and im failing to find what its mising maybe i should post the code here?

**Reply by pmconne** [reply]
https://www.npmjs.com/package/@itwin/imodels-access-backend

**Reply by Nhillus** [reply]
could you point me to a doc that uses a basic example of briefcase or the general flow to use a briefcase, or maybe a repo that use it  Thank you again!

**Reply by pmconne** [reply]
Maybe @aruniverse can.

**Reply by aruniverse** [reply]
Hey! We're excited to hear of your interest in using iTwin.js and look forward to hearing more feedback.

Have you looked at starting off with the [Desktop Viewer Template](https://github.com/iTwin/viewer/tree/master/packages/templates/cra-template-desktop-viewer) to get started? This is a wrapper for many of the iTwin.js Core APIs and should make it easier for you to get started. The template is geared towards `Read` only use cases, but by changing a few lines you can get it to actually `Write` back to the iModel as well.

You can see [here](https://github.com/iTwin/viewer/blob/master/packages/templates/cra-template-desktop-viewer/template/src/backend/main.ts#L42) where we provide an instance of BackendHubAccess to IModelHost. 

Code Changes required to write back to iModel:

- [src/frontend/hooks/useDownload.tsx](https://github.com/iTwin/viewer/blob/master/packages/templates/cra-template-desktop-viewer/template/src/frontend/hooks/useDownload.tsx#L45) 
		- modify [SyncMode](https://www.itwinjs.org/reference/core-common/imodels/syncmode/) in [NativeApp.requestDownloadBriefcase](https://www.itwinjs.org/reference/core-frontend/nativeapp/downloadbriefcaseoptions/) from `PullOnly` to `PullAndPush`

- [src/frontend/components/modelSelector/SelectIModel.tsx](https://github.com/iTwin/viewer/blob/master/packages/templates/cra-template-desktop-viewer/template/src/frontend/components/modelSelector/SelectIModel.tsx#L60)
	- change `readonly` from `false` to `true` to download a briefcase that you can write to

- [src/frontend/components/routes/ViewerRoute.tsx](https://github.com/iTwin/viewer/blob/master/packages/templates/cra-template-desktop-viewer/template/src/frontend/components/routes/ViewerRoute.tsx#L42)
	- add the [readonly prop](https://github.com/iTwin/viewer/blob/7189dae2f67d2eeb08d7ec51e747d1a3c86a12b0/packages/modules/viewer-react/src/types.ts#L163), and set it to false so you once you open the iModel, you can actually write to it

**Reply by Nhillus** [reply]
@aruniverse  Thank you im checking out the repos that you mention!

**Comment by ColinKerr** [comment]
You can use ECSql to query for the schemas used to store data in the iModel.  It's exposed via the ECDbMeta schema.

Here is some examples from our doc site: https://www.itwinjs.org/learning/ecsqltutorial/metaqueries/

I hope this helps.  If you have more specific questions about querying for EC (db) schemas in an iModel.
